data_model/src/lib.rs

```diff
@@ -535,12 +646,14 @@ pub trait ValueMarker: Debug + Clone + Into<Value> {}
 impl<V: Into<Value> + Debug + Clone> ValueMarker for V {}
 
 /// This trait marks entity that implement it as identifiable with an `Id` type to find them by.
-pub trait Identifiable: Debug + Clone {
-    /// Defines the type of entity's identification.
+pub trait Identifiable: Debug + Clone + From<Self::Constructor> {
+    /// Type of entity's identification.
     type Id: Into<IdBox> + Debug + Clone + Eq + Ord;
+    /// Type of constructor of the entity
+    type Constructor: Into<RegistrableBox>;
 }
```

Comment:
```suggestion
/// Can be identified in the blockchain.
pub trait Identifiable: Debug + Clone {
    /// Identification.
    type Id: Into<IdBox> + Debug + Clone + Eq + Ord;
    /// The first appearance on the blockchain.
    type RegisterSet: Registrable;
}

trait Registrable: Into<RegistrableBox> {
    type Object: Identifiable;
    fn new(id: <Self::Object as Identifiable>::Id) -> Self;
    fn build(self) -> Self::Object;
}
```

Reply:
ok, @Arjentix had a similar suggestion. We can go with something like this

Reply:
Removed from my suggestion `Identifiable::RegisterSet`, which may cause a conflict with `Registrable::Object`

Reply:
basically, you'd like to have a `build` method? This is why you're proposing the change?

Reply:
Just to mention, AFAIK Rust doesn't like to use *abble* suffix in traits: https://rust-lang.github.io/rfcs/0344-conventions-galore.html#trait-naming

Reply:
I'd like to keep `Identifiable` as is with `RegisteredWith` associated type. Because then I can reason about `Identifiable`s not mentioning types they depend on for registration. E.g. I can write in `core/smartcontracts/isi/world.rs`:
```rs
    impl<W: WorldTrait> Execute<W> for Register<Domain>;
```
vs.
```rs
    impl<W: WorldTrait> Execute<W> for Register<NewDomain>;
```

If you want a builder I can propose a separate `Registrable` trait like:
```rs
/// This trait marks entity that implement it as identifiable with an `Id` type to find them by.
pub trait Identifiable: Debug + Clone + From<Self::RegisteredWith> {
    /// Type of entity's identification.
    type Id: Into<IdBox> + Debug + Clone + Eq + Ord;
    /// Type used to register `Identifiable` entity
    type RegisteredWith: Registrable<Registers = Self>;
}

trait Registrable: Into<RegistrableBox> {
    type Registers: Identifiable<RegisteredWith = Self>;
    fn build(self) -> Self::Registers;
}
```

Reply:
> Just to mention, AFAIK Rust doesn't like to use _abble_ suffix in traits: [rust-lang.github.io/rfcs/0344-conventions-galore.html#trait-naming](https://rust-lang.github.io/rfcs/0344-conventions-galore.html#trait-naming)

@Arjentix I am aware this is a slight deviation from the Rust convention, but we've already had `Identifiable`

Reply:
> I'd like to keep `Identifiable` as is with `RegisteredWith` associated type. Because then I can reason about `Identifiable`s not mentioning types they depend on for registration. E.g. I can write in `core/smartcontracts/isi/world.rs`:
> 
> ```rust
> impl<W: WorldTrait> Execute<W> for Register<Domain>;
> ```
> 
> vs.
> 
> ```rust
> impl<W: WorldTrait> Execute<W> for Register<NewDomain>;
> ```
> 
> If you want a builder I can propose a separate `Registrable` trait like:
> 
> ```rust
> /// This trait marks entity that implement it as identifiable with an `Id` type to find them by.
> pub trait Identifiable: Debug + Clone + From<Self::RegisteredWith> {
>     /// Type of entity's identification.
>     type Id: Into<IdBox> + Debug + Clone + Eq + Ord;
>     /// Type used to register `Identifiable` entity
>     type RegisteredWith: Registrable<Registers = Self>;
> }
> 
> trait Registrable: Into<RegistrableBox> {
>     type Registers: Identifiable<RegisteredWith = Self>;
>     fn build(self) -> Self::Registers;
> }
> ```

Then do wee need `From<Self::RegisteredWith>` as `Identifiable` restriction? I think no

Reply:
Okay, either
```rust
trait Identifiable: From<Self::RegisteredWith>
```
or `Registrable::build` is fine to me

Reply:
BTW am I the only one who senses a passive nuance in `trait Doable`?

Reply:
@s8sato don't understand what you mean

Reply:
I suppose a pattern like
```rust
trait Do {
    type Obj: Doable;
    fn do(self, _obj: Self::Obj) {}
}
struct Subject;
impl Do for Subject {
    type Obj = Object;
}

trait Doable {}
struct Object;
impl Doable for Object {}

fn main() {
    Subject.do(Object)
}
```

Reply:
I wouldn't generalize at this point. I've added `build` methods to `NewDomain` and `NewAccount` without introducing the `Registrable` trait. The trait seems redundant and it would have to be included at the call site where `.build()` method is used just to be able to use the method

Can we resolve this conversation?

Reply:
I don't think we need to have both of [these](https://github.com/hyperledger/iroha/pull/1984#discussion_r834229688)

Reply:
Ah, I meant to say not `Registrable::build` but `NewDomain::build` and `NewAccount::build`.
So I think you can just reject the suggestion and resolve this conversation